### PR TITLE
Fix stalled method on DataSource; tests to prove it is fixed

### PR DESCRIPTION
**Pull request by eanders**
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

## Description
[//]: # (Summarize changes and include links related issue. List any new dependencies)
* Fixes calculation to determine if a data source that is setup to auto import is stalled
* Tests to confirm edge cases

## Type of change
[//]: # (e.g., Bug fix, New feature, Code clean-up, Dependency update)

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] I have updated the documentation (or not applicable)
- [x] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

[//]: # NOTE: system tests may fail if there is no branch on the hmis-frontend that matches the Source  or Target branch of this PR. This is expected

### CODE REVIEW

File: app/models/grda_warehouse/data_source.rb [hunk 1]
```diff
@@ -497,36 +497,40 @@ def unprocessed_enrollment_count
...
   end
 
-  # Have we received the expected number of files at least once over the past 48 hours?
-  # there is an assumption that within
-  # return the date of the most-recent fully successful import
-  def stalled_since?(date)
-    return nil unless date.present?
+  # Returns the date of the most recent fully successful import if the import is stalled, nil if it is not stalled
+  # @return [Date, nil] The date the import stalled, or nil if not stalled.
+  def stalled?
```
Comment (ttoomey) [inline app/models/grda_warehouse/data_source.rb]: Suggest removing the '?' since it's not returning a boolean
 ```suggestion
   def stall_date
```

File: app/models/grda_warehouse/data_source.rb [hunk 2]
```diff
@@ -497,36 +497,40 @@ def unprocessed_enrollment_count
...
+      # and it came in the last 24 hours, we're good
+      return nil if min_completion_time > 24.hours.ago
 
-    most_recent_upload.completed_at.to_date
+      # if not, return the last time we received a file
+      return min_completion_time.to_date
+    end
+
+    # If we processed the expected number of files within a 24 hour period, we're good
+    return nil if min_completion_time > 24.hours.ago && received_files_count == hmis_import_config.file_count
```
Comment (ttoomey) [inline app/models/grda_warehouse/data_source.rb]: looks like this would appear as stalled if we processed more than the expected number... is that what we want?
```suggestion
    return nil if min_completion_time > 24.hours.ago && received_files_count >= hmis_import_config.file_count
```

File: app/models/grda_warehouse/data_source.rb [hunk 3]
```diff
@@ -497,36 +497,40 @@ def unprocessed_enrollment_count
...
 
     # hmis_import_config.file_count is the expected number of uploads for a given day
     # fetch the expected number, and confirm they all arrived within a 24 hour window
     most_recent_uploads = uploads.completed.
-      # limit look back to 6 months to improve performance
-      where(user_id: User.system_user.id, completed_at: 6.months.ago.to_date..Date.current).
-      order(created_at: :desc).
+      # limit look back to 6 months to improve performance, but to potentially highlight missing data
+      where(user_id: User.system_user.id, completed_at: 6.months.ago..DateTime.current).
+      order(completed_at: :desc, created_at: :desc).
```
Comment (ttoomey) [inline app/models/grda_warehouse/data_source.rb:511]: We assume unique files here right? If we expected two different files but just got one of the two files twice, is that a problem? Wondering if there should be count distinct on upload.file

File: app/models/grda_warehouse/data_source.rb [hunk 4]
```diff
@@ -497,36 +497,40 @@ def unprocessed_enrollment_count
...
     return nil unless hmis_import_config&.active
 
     # hmis_import_config.file_count is the expected number of uploads for a given day
     # fetch the expected number, and confirm they all arrived within a 24 hour window
     most_recent_uploads = uploads.completed.
-      # limit look back to 6 months to improve performance
-      where(user_id: User.system_user.id, completed_at: 6.months.ago.to_date..Date.current).
-      order(created_at: :desc).
+      # limit look back to 6 months to improve performance, but to potentially highlight missing data
+      where(user_id: User.system_user.id, completed_at: 6.months.ago..DateTime.current).
```
Comment (ttoomey) [inline app/models/grda_warehouse/data_source.rb]: probably favor Time or DateTime. Also do we need to exclude dates in the future?
```suggestion
      where(user_id: User.system_user.id, completed_at: 6.months.ago..Time.current).
```

**Review by ttoomey — APPROVED** [review]
looks good, I left a few suggestions

File: app/models/grda_warehouse/data_source.rb (hunk 2 above)
Comment (eanders) [reply app/models/grda_warehouse/data_source.rb]: That _should_ be impossible since we're limiting to `first(hmis_import_config.file_count)` in the query, but no harm.

File: app/models/grda_warehouse/data_source.rb (hunk 3 above)
Comment (eanders) [reply app/models/grda_warehouse/data_source.rb:511]: Not sure how you could get duplicates given that it is just querying one table, but no harm in a distinct since we're selecting a limited set of columns so it won't try to distinct on a json column.
